core/src/main/scala/org/broadinstitute/dsde/rawls/billing/BpmBillingProjectLifecycle.scala

```diff
@@ -75,45 +80,130 @@ class BpmBillingProjectLifecycle(
                                  config: MultiCloudWorkspaceConfig,
                                  ctx: RawlsRequestContext
   ): Future[CreationStatus] = {
-    val projectName = createProjectRequest.projectName.value
-    try {
-      val profileModel = blocking {
-        billingProfileManagerDAO.createBillingProfile(projectName, createProjectRequest.billingInfo, ctx)
-      }
+    val projectName = createProjectRequest.projectName
+
+    def createBillingProfile: Future[ProfileModel] =
+      Future(blocking {
+        val profileModel = billingProfileManagerDAO.createBillingProfile(
+          projectName.value,
+          createProjectRequest.billingInfo,
+          ctx
+        )
+        logger.info(
+          s"Creating BPM-backed billing project ${projectName.value}, created profile with ID ${profileModel.getId}."
+        )
+        profileModel
+      })
 
-      val landingZoneResponse = blocking {
-        // This starts a landing zone creation job. There is a separate monitor that polls to see when it
-        // completes and then updates the billing project status accordingly.
+    // This starts a landing zone creation job. There is a separate monitor that polls to see when it
+    // completes and then updates the billing project status accordingly.
+    def createLandingZone(profileModel: ProfileModel): Future[CreateLandingZoneResult] =
+      Future(blocking {
         workspaceManagerDAO.createLandingZone(
           config.azureConfig.get.landingZoneDefinition,
           config.azureConfig.get.landingZoneVersion,
           profileModel.getId,
           ctx
         )
-      }
-      Option(landingZoneResponse.getErrorReport) match {
-        case Some(errorReport) =>
-          throw new LandingZoneCreationException(
-            RawlsErrorReport(StatusCode.int2StatusCode(errorReport.getStatusCode), errorReport.getMessage)
-          )
-        case None =>
-          logger.info(
-            s"Creating BPM-backed billing project ${projectName}, initiated creation of landing zone ${landingZoneResponse.getLandingZoneId} with jobId ${landingZoneResponse.getJobReport.getId}"
-          )
-      }
-      for {
-        _ <- billingRepository.updateLandingZoneId(createProjectRequest.projectName,
-                                                   landingZoneResponse.getLandingZoneId
+      })
+
+    createBillingProfile.flatMap { profileModel =>
+      createLandingZone(profileModel)
+        .flatMap { landingZone =>
+          (for {
+            _ <- Option(landingZone.getErrorReport).traverse { errorReport =>
+              Future.failed(
+                new LandingZoneCreationException(
+                  RawlsErrorReport(StatusCode.int2StatusCode(errorReport.getStatusCode), errorReport.getMessage)
+                )
+              )
+            }
+            _ <- Option(landingZone.getJobReport) match {
```

Comment:
@ehigham I agree we should avoid a potential NullPointerException. Is there a way to do this that is like `traverse` but only gives you the None case?

Reply:
`getOrElse` would make you handle the `None` case explicitly I guess?